Source/JavaScriptCore/runtime/WaiterListManager.cpp

```diff
@@ -0,0 +1,256 @@
+
+            while (waiter.getVM() && time.nowWithSameClock() < time)
+                waiter.getCondition().waitUntil(list->lock, time.approximateWallTime());
+            ASSERT(!waiter.getVM() || waiter.getVM() == &vm);
+            bool didGetDequeued = !waiter.getVM();
+
+            if (didGetDequeued)
+                return vm.smallStrings.okString();
+            list->take(locker, &waiter);
+            return vm.smallStrings.timedOutString();
+        }
+    } else
+        return vm.smallStrings.notEqualString();
+}
+
+void WaiterListManager::addAsyncWaiter(void* ptr, JSPromise* promise, Seconds timeout)
+{
+    WaiterList* list = findOrCreateList(ptr);
+    Ref<Waiter> waiter = adoptRef(*new Waiter(promise));
+
+    {
+        Locker locker { list->lock };
+        list->addLast(locker, waiter.ptr());
+
+        if (timeout != Seconds::infinity()) {
+            Ref<RunLoop::DispatchTimer> timer = RunLoop::current().dispatchAfter(timeout, [this, ptr, waiter = waiter.copyRef()] {
+                WaiterList* list = findList(ptr);
+                if (list) {
+                    Locker locker { list->lock };
+                    notifyWaiterImpl(locker, list->take(locker, waiter.ptr()), ResolveResult::Timeout);
+                }
+            });
+            waiter->setTimer(WTFMove(timer));
+        }
+    }
+
+    dataLogLnIf(WaiterListsManagerInternal::verbose, "WaiterListManager added a new AsyncWaiter ", RawPointer(waiter.ptr()), " to a waiterList for ptr ", RawPointer(ptr));
+}
+
+unsigned WaiterListManager::notifyWaiter(void* ptr, unsigned count)
+{
+    ASSERT(ptr);
+    unsigned notified = 0;
+    WaiterList* list = findList(ptr);
+    if (list) {
+        Locker locker { list->lock };
+        while (notified < count && list->size()) {
+            notifyWaiterImpl(locker, list->takeFirst(locker), ResolveResult::Ok);
+            notified++;
+        }
+    }
+
+    dataLogLnIf(WaiterListsManagerInternal::verbose, "WaiterListManager notified waiters (count ", notified, ") for ptr ", RawPointer(ptr));
+    return notified;
+}
+
+void WaiterListManager::notifyWaiterImpl(const AbstractLocker&, Waiter* waiter, const ResolveResult resolveResult)
+{
+    ASSERT(waiter);
+
+    if (waiter->isAsync()) {
+        VM& vm = *waiter->getVM();
+        auto ticket = waiter->getTicket();
+        vm.deferredWorkTimer->scheduleWorkSoon(ticket, [resolveResult](DeferredWorkTimer::Ticket ticket) {
+            JSPromise* promise = jsCast<JSPromise*>(ticket->target());
+            JSGlobalObject* globalObject = promise->globalObject();
+            VM& vm = promise->vm();
+            JSValue result = resolveResult == ResolveResult::Ok ? vm.smallStrings.okString() : vm.smallStrings.timedOutString();
+            promise->resolve(globalObject, result);
+        });
+
+        if (resolveResult == ResolveResult::Ok)
+            waiter->cancelTimer();
```

Comment:
is it safe?

Reply:
Yes, I believe so. The `timer` is created and added to `RunLoop` with `CFRunLoopAddTimer` initially. Then `cancelTimer` will stop the `timer` with `CFRunLoopTimerInvalidate`. Since `waiterList` is locked only one thread can access remove the waiter and call `cancelTimer`.